server-kotlin/src/main/kotlin/digital/cesko/internet_stream/InternetStreamService.kt

```diff
@@ -0,0 +1,200 @@
+
+@Service
+class InternetStreamService() {
+    val csvFormat = CSVFormat.DEFAULT.withDelimiter(';').withFirstRecordAsHeader()
+            .withIgnoreHeaderCase().withTrim()
+
+    @Value("\${budgets.baseSourceUrls}") val baseUrls: MutableList<String> = mutableListOf()
+    @Value("\${budgets.fileUrls}") val fileUrls: MutableList<String> = mutableListOf()
+    @Value("\${budgets.cityUrls}") val cityUrls: MutableList<String> = mutableListOf()
+    val bufferSize = 8192
+
+    @Scheduled(fixedRateString = "\${budgets.frequency}")
+    fun fetchData() {
+        val profileIds = getProfilesIds()
+
+        for ((index,baseUrl) in baseUrls.withIndex()) {
+            val completeBudgetPerCity: MutableList<Budget> = mutableListOf()
+            val cityId = profileIds[index]
+            for (fileUrl in fileUrls) {
+                var zipInputStream: ZipInputStream? = null
+                var downloadInputStream: InputStream? = null
+                val budgetPerYear: MutableList<Budget> = mutableListOf()
+                var url: String? = null
+                try {
+                    url = baseUrl + fileUrl
+                    val downloadUrl = URL(url)
+                    downloadInputStream = downloadUrl.openStream()
+                    zipInputStream = ZipInputStream(downloadInputStream)
+                    var ze: ZipEntry? = zipInputStream.nextEntry
+                    while (ze != null) {
+                        if (ze.name == "RU.csv" || ze.name == "SK.csv") {
+                            val byteArray = unzipFromStream(zipInputStream)
+                            val budget = parseBudgets(byteArray)
+                            budgetPerYear.addAll(budget)
+                        }
+                        ze = zipInputStream.nextEntry
+                    }
+                    completeBudgetPerCity.addAll(budgetPerYear)
+                } catch (malformedUrl: MalformedURLException) {
+                    logger.error("Unable to download budgets from $url", malformedUrl)
+                } catch (ioException: IOException) {
+                    logger.error("Unable to process budgets from $url", ioException)
+                } finally {
+                    zipInputStream?.close()
+                    downloadInputStream?.close()
+                }
+            }
+            deleteCityBudgets(cityId)
+            saveCityBudgets(cityId, completeBudgetPerCity)
+            completeBudgetPerCity.clear()
+        }
+    }
+
+    // Get profile_id of every city defined in application.properties.
+    // Search is made by column url.
+    fun getProfilesIds(): MutableList<Int> {
+        val profileIds: MutableList<Int> = mutableListOf()
+        for (cityUrl in cityUrls) {
+            val resultRow: CityExport = transaction {
+                Profiles.select { Profiles.url eq cityUrl }
+                        .map { toProfileCityExport(it) }
+                        .ifEmpty { throw Exception("CityUrl $cityUrl not found.") }
```

Comment:
Co se ma stat pri tehle vyjimce? Tak jak je to napsane to tu vyjimku vyhodi scheduleru, coz asi neni idealni.